### PR TITLE
Refactor mongodb interface implementation

### PR DESCRIPTION
**Pull request by idegtiarov**

### CODE REVIEW

File: edx_adapt/data/course_repository.py [hunk 1]
```diff
@@ -516,6 +236,28 @@ def advance_problem(self, course_id, user_id):
...
 
+    def get_all_remaining_problems(self, course_id, user_id):
+        return [x for x in self._get_remaining_by_user(course_id, user_id)
+                if x['pretest'] is False and x['posttest'] is False]
+
+    def get_remaining_problems(self, course_id, skill_name, user_id):
+        remaining = self.get_all_remaining_problems(course_id, user_id)
+        return [x for x in remaining if skill_name in x['skills']]
+
+    def get_all_remaining_posttest_problems(self, course_id, user_id):
```
Comment (z4y4ts) [inline edx_adapt/data/course_repository.py]: It's not quite clear what's the difference between this method and `get_remaining_posttest_problems`?

File: edx_adapt/data/mongodb_storage.py [hunk 2]
```diff
@@ -1,70 +1,245 @@
...
+        return [course.get('course_id') for course in self.db.Courses.find()]
+
+    def get(self, coll_name, key):
+        """
+        Returns value for the required key from db[coll_name]
+
+        :param coll_name: name of collection
+        :param key: key which value is return
+        """
+        # FIXME(idegtiarov) Such structure is still been using in Generic coll_name, it would be great to refactor it
```
Comment (z4y4ts) [inline edx_adapt/data/mongodb_storage.py]: been using -> being used, If I understand you correctly

File: edx_adapt/data/mongodb_storage.py [hunk 3]
```diff
@@ -1,70 +1,245 @@
...
+
+    def course_user_done(self, course_id, user_id):
+        """
+        Move student to field from 'users_in_progress' to 'users_finished'
+        """
+        update_params = [{'course_id': course_id}, {'$pull': {'users_in_progress': user_id},
+                                                    '$push': {'users_finished': user_id}}]
+        self.db.Courses.update_one(*update_params)
+
+    def get_statistic(self, collection, user_id, filter_condition, group_key, group_id=None, op='$sum', op_value=1):
```
Comment (z4y4ts) [inline edx_adapt/data/mongodb_storage.py]: get_statistics

File: edx_adapt/tests/logic_tests.py [hunk 4]
```diff
@@ -24,18 +46,21 @@ class BaseTestCase(unittest.TestCase):
...
 
     @classmethod
     def tearDownClass(cls):
         # NOTE(idegtiarov) sqlite is too slow for using on server we will support only MongoDB
-        mclient = pymongo.MongoClient()['edx-adapt']
-        mclient.drop_collection(cls.course_id)
-        regex = re.compile('_student_')
-        mclient.Generic.remove({'key': {'$regex': regex}})
+        mclient = pymongo.MongoClient()
+        mclient.drop_database('edx-adapt')
```
Comment (z4y4ts) [inline edx_adapt/tests/logic_tests.py]: It's better to name test db 'edx-adapt-test' or something like that. Just to not drop real db by an accident.

File: edx_adapt/tests/logic_tests.py [hunk 5]
```diff
@@ -1,20 +1,42 @@
...
 base_api_path = '/api/v1/course'
 
 
+def _setup_course_in_edxadapt(client, **kwargs):
+    client.post(base_api_path, data=json.dumps({'course_id': kwargs['course_id']}), headers=kwargs['headers'])
+    for skill in kwargs['skills']:
+        payload = json.dumps({'skill_name': skill})
+        client.post(base_api_path + '/{course_id}/skill'.format(**kwargs), data=payload, headers=kwargs['headers'])
+    path_to_file = os.path.join(os.path.dirname(__file__), '../../data/BKT/problems.csv')
+    for row in [line.strip().split(',') for line in open(path_to_file).readlines()]:
```
Comment (z4y4ts) [inline edx_adapt/tests/logic_tests.py]: Use `csv.DictReader`

File: edx_adapt/data/course_repository.py (hunk 1 above)
Comment (idegtiarov) [inline edx_adapt/data/course_repository.py]: get_remaining_posttest_problems returns posttest problem for the certain skill which is set as parameter

File: edx_adapt/data/mongodb_storage.py (hunk 2 above)
Comment (idegtiarov) [inline edx_adapt/data/mongodb_storage.py]: done

File: edx_adapt/data/mongodb_storage.py (hunk 3 above)
Comment (idegtiarov) [inline edx_adapt/data/mongodb_storage.py]: done

File: edx_adapt/tests/logic_tests.py (hunk 5 above)
Comment (idegtiarov) [inline edx_adapt/tests/logic_tests.py]: we do not have column titles in csv files, and csv.DictReader won't work as expected right now

File: edx_adapt/tests/logic_tests.py (hunk 4 above)
Comment (idegtiarov) [inline edx_adapt/tests/logic_tests.py]: Actually tests are run on local database not on adapt server and BTW we have a lot of hard code now with database initializing when application is started up.

File: edx_adapt/tests/logic_tests.py (hunk 5 above)
Comment (z4y4ts) [inline edx_adapt/tests/logic_tests.py]: Ok, then `csv.reader()`

File: edx_adapt/tests/logic_tests.py (hunk 4 above)
Comment (z4y4ts) [inline edx_adapt/tests/logic_tests.py]: Yes, but then you have a dev db on a local machine.
I just don't want to add a code which can let some stupid error or typo ruin data. Even if it's only on a dev machine.

If there's a lot of hardcode needs to be rewritten to make this change it's ok to do it in a separate PR, but put a TODO note here.

File: edx_adapt/data/course_repository.py (hunk 1 above)
Comment (idegtiarov) [inline edx_adapt/data/course_repository.py]: make sense, done

File: edx_adapt/tests/logic_tests.py (hunk 5 above)
Comment (idegtiarov) [inline edx_adapt/tests/logic_tests.py]: make sense, done

File: edx_adapt/tests/logic_tests.py (hunk 4 above)
Comment (idegtiarov) [inline edx_adapt/tests/logic_tests.py]: agreed, will add TODO flag

File: edx_adapt/tests/logic_tests.py [hunk 6]
```diff
@@ -19,18 +20,19 @@ def _setup_course_in_edxadapt(client, **kwargs):
...
-            base_api_path + '/{course_id}'.format(**kwargs), data=payload, headers=kwargs['headers']
-        )
+    with open(path_to_file) as file_csv:
+        for row in csv.reader(file_csv):
+            problem = row[0]
+            skill = row[1]
+            url = 'http://edx-lms.raccoongang.com/courses/{course_id}/{pname}'.format(pname=problem, **kwargs)
+            pre = "Pre_a" in problem
+            post = "Post_a" in problem
+            payload = json.dumps({
```
Comment (z4y4ts) [inline edx_adapt/tests/logic_tests.py]: No need to encode into json manually. [Requests can do this for you](http://docs.python-requests.org/en/master/user/quickstart/#more-complicated-post-requests).

Comment (idegtiarov) [inline edx_adapt/tests/logic_tests.py]: In tests Flask.test_client is used and it expects that json is encoded manually.
